sys/net/gnrc/network_layer/sixlowpan/frag/gnrc_sixlowpan_frag.c

```diff
@@ -212,51 +216,67 @@ static uint16_t _send_nth_fragment(gnrc_sixlowpan_netif_t *iface, gnrc_pktsnip_t
+    size_t payload_len = gnrc_pkt_len(fragment_msg->pkt->next);
+    msg_t msg;
 
 #if defined(DEVELHELP) && defined(ENABLE_DEBUG)
     if (iface == NULL) {
         DEBUG("6lo frag: iface == NULL, expect segmentation fault.\n");
-        gnrc_pktbuf_release(pkt);
+        gnrc_pktbuf_release(fragment_msg->pkt);
         return;
     }
 #endif
 
-    if ((res = _send_1st_fragment(iface, pkt, payload_len, datagram_size)) == 0) {
-        /* error sending first fragment */
-        DEBUG("6lo frag: error sending 1st fragment\n");
-        gnrc_pktbuf_release(pkt);
-        return;
-    }
-
-    offset += res;
-    thread_yield();
-
-    /* (offset + (datagram_size - payload_len) < datagram_size) simplified */
-    while (offset < payload_len) {
-        if ((res = _send_nth_fragment(iface, pkt, payload_len, datagram_size,
-                                      offset)) == 0) {
-            /* error sending subsequent fragment */
-            DEBUG("6lo frag: error sending subsequent fragment (offset = %" PRIu16
-                  ")\n", offset);
-            gnrc_pktbuf_release(pkt);
+    /* Check weater to send the first or an Nth fragment */
+    if (fragment_msg->offset == 0) {
+        /* increment tag for successive, fragmented datagrams */
+        _tag++;
+        if ((res = _send_1st_fragment(iface, fragment_msg->pkt, payload_len, fragment_msg->datagram_size)) == 0) {
+            /* error sending first fragment */
+            DEBUG("6lo frag: error sending 1st fragment\n");
+            gnrc_pktbuf_release(fragment_msg->pkt);
             return;
         }
+        fragment_msg->offset += res;
 
-        offset += res;
+        /* send message to self*/
+        msg.type = GNRC_SIXLOWPAN_MSG_FRAG_SND;
+        msg.content.ptr = (void *)fragment_msg;
+        msg_send_to_self(&msg);
         thread_yield();
     }
-
-    /* remove original packet from packet buffer */
-    gnrc_pktbuf_release(pkt);
-    _tag++;
+    else {
+        /* (offset + (datagram_size - payload_len) < datagram_size) simplified */
+        if (fragment_msg->offset < payload_len) {
+            if ((res = _send_nth_fragment(iface, fragment_msg->pkt, payload_len, fragment_msg->datagram_size,
+                                          fragment_msg->offset)) == 0) {
+                /* error sending subsequent fragment */
+                DEBUG("6lo frag: error sending subsequent fragment (offset = %" PRIu16
+                      ")\n", fragment_msg->offset);
+                gnrc_pktbuf_release(fragment_msg->pkt);
+                return;
+                }
+            fragment_msg->offset += res;
+
+            /* send message to self*/
+            msg.type = GNRC_SIXLOWPAN_MSG_FRAG_SND;
+            msg.content.ptr = (void *)fragment_msg;
+            msg_send_to_self(&msg);
+            thread_yield();
+        }
+        else {
+            /* Set state to idle, free for next framentation */
+            fragment_msg->state = GNRC_SIXLOWPAN_FRAG_STATE_IDLE;
```

Comment:
Also, since `GNRC_SIXLOWPAN_FRAG_STATE_IDLE` is non-zero, the state won't be initialized to IDLE on start-up, unless you set it somewhere to that value... All-in-all, using `fragment_msg->pkt` as state seems far more convenient to me ;-)